craft_providers/bases/instance_config.py

```diff
@@ -0,0 +1,118 @@
+# Copyright 2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Persistent instance config / datastore resident in provided environment."""
+
+import dataclasses
+import io
+import logging
+import pathlib
+import subprocess
+from typing import TYPE_CHECKING, Dict, Optional
+
+import pydantic
+import yaml
+
+from craft_providers import Executor, errors
+
+from .errors import BaseConfigurationError
+
+# Avoid pyright errors by using dataclasses.dataclass during type checking by
+# importing the standard library version.
+# https://github.com/microsoft/pyright/issues/637
+if TYPE_CHECKING:
+    from dataclasses import dataclass
+else:
+    from pydantic.dataclasses import dataclass
```

Comment:
What happens if you type check with pydantic dataclass?

Reply:
IIRC pyright gives some errors. like `error: Missing named argument` that's simply wrong.

Reply:
https://github.com/microsoft/pyright/issues/637